applications/asset_tracker_v2/CMakeLists.txt

```diff
@@ -6,6 +6,11 @@
 
 cmake_minimum_required(VERSION 3.20.0)
 
+if(NOT DEFINED OVERLAY_CONFIG)
+  message(WARNING "Overlay configuration list is empty, using nRF Cloud as the default cloud provider")
+  list(APPEND OVERLAY_CONFIG "overlay-nrf_cloud.conf")
```

Comment:
we should not apply any config fragments per default.

Users can create a custom `my_conf.conf` and use it with a sample like: `cmake ... -DCONF_FILE=/<path>/my_conf.conf` and then build the sample with known set of settings.

But in this app you would apply an additional config fragment begind user's back in such case.

That is not acceptable.

Reply:
Do you have a proposal on how to solve this? We want the user to have a functional sample when building without options.

Reply:
> Board-specific config files are also patched in behind the user's "back".
> How is that different?

Actually board specific config files are not patched behind users back.

The board fragments is a documented feature, and it only applies when default `prf.conf` files are used:
https://docs.zephyrproject.org/latest/guides/build/kconfig/setting.html#the-initial-configuration
Or a `prj_<build>.conf` is in use, in which case the `boards/<board>_<build>.conf` is used, if exists.

Test for yourself:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/subsys/settings
...
Parsing /projects/github/ncs/zephyr/Kconfig
Loaded configuration '/projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration '/projects/github/ncs/zephyr/samples/subsys/settings/prj.conf'
Merged configuration '/projects/github/ncs/zephyr/samples/subsys/settings/boards/nrf52840dk_nrf52840.conf'
Configuration saved to '/projects/github/ncs/zephyr/build/zephyr/.config'
```

but when using a custom conf file, the board file is NOT used:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/subsys/settings -DCONF_FILE=myconftest.conf
...
Parsing /projects/github/ncs/zephyr/Kconfig
Loaded configuration '/projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration 'myconftest.conf'
Configuration saved to '/projects/github/ncs/zephyr/build/zephyr/.config'
```
Notice, that with `-DCONF_FILE=myconftest.conf` then `boards/nrf52840dk_nrf52840.conf'` is no longer used.

> Isn't that sufficient enough?

you are still deviating on how things are generally working.
We can't expect users to read each and every user guide in details to understand exactly how this sample decides to add extra configurations. And then another sample does things a bit differently.

Secondly, user's cannot avoid `overlay-nrf_cloud.conf` to be applied with this PR, unless a dummy / empty fragment is created and applied.

Reply:
> o you have a proposal on how to solve this? We want the user to have a functional application when building without options. And at the same time we do not want them to comment out several config options if they want to build with another cloud vendor backend. (nrfcloud(default), aws and azure))

Let me take a closer look at the sample.
One question, is it allowed to select multiple cloud vendor backends, like both nrfcloud and aws, or are they mutual exclusive ?

Reply:
Mutual exclusive.

Reply:
Alright, thanks for the explanation @tejlmand. Yes, they are mutual exclusive. The application requires that either `overlay-nrf_cloud.conf`, `overlay-azure.conf`, or `overlay-aws.conf` be configured with the application. The rest is the overlays are optional.

Reply:
IMHO the originally proposed PR is acceptable.

But if it must be changed such that OVERLAY_CONFIG is not set by the app build scripts then we could do

if(NOT DEFINED CONF_FILE)
set(CONF_FILE prj.conf overlay-nrf_cloud.conf)
endif()

and instruct the user to use a list of CONF_FILE instead of OVERLAY_CONFIG.

Is this acceptable @tejlmand, or do you have a better suggestion perhaps ?

Reply:
> However, the nRF Cloud library suite has other means of communications with nRF Cloud such as via REST. Then CONFIG_NRF_CLOUD_MQTT isn't needed.

Yes, but a default doesn't mean it can't be disabled.
Of course users not wanting to use `MQTT` would then have to opt-out, like:
```
CONFIG_NRF_CLOUD=y
CONFIG_NRF_CLOUD_MQTT=n
```
but if `MQTT` is a good default it will allow majority of samples to simply do:
```
CONFIG_NRF_CLOUD=y
```
and forget about `MQTT`.
Also reducing number of lines in our cloud sample config files.

I'll leave the decision on this for you guys, just mentioning my observations.

> If this option is enabled, the socket is exposed and its expected that the polling is done in the application.

Did you perhaps mean `If this option is disabled` ?

> >  Same regarding the setting: CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD=y.
>
> This option can probably default to y. AFAIK there are no apps/samples that uses nRF Cloud MQTT that polls in the application.

It sounds to me this should have a `default y`.
Remember, samples can always disable this with `CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD=n`, and sane default values make life easier for majority of users, and reduces number of lines in our config files.

> so why not have it defaulting to GPS_MODULE_NMEA when NRF_CLOUD is selected ?
>
> Originally, the idea was to keep the GPS module as generic as possible so that it could be easily be reused in other applications and not have artifacts from Kconfigs options / functionality from non-related libraries.

In that case you should give the choice a variable name, as that will allow to set the default in another Kconfig file, like:
```
choice GPS_MODULE_DATA_FORMAT
	default GPS_MODULE_NMEA if NRF_CLOUD
endchoice
```
and then in the current Kconfig code for the GPS, you would have:
```
choice GPS_MODULE_DATA_FORMAT
	prompt "Select GNSS data format"
	default GPS_MODULE_PVT
...
endchoice
```
That way you may keep knowledge where it rightfully belongs.

Somewhere I assume there must be a dependency.
Either GPS knows about cloud service (probably not).
Or a cloud service knows something about GPS (it seems nRF Cloud has some sort of expectation on GPS for positioning feature)
Or a dedicated Kconfig that glues such knowledge together.

As long as the GPS module is application specific, the `<app>/Kconfig` file sounds like the right location for this.
When / if GPS module is taken out as a generic module, then it can be considered where such code should be placed.